### PR TITLE
[SPARK-6612] [MLLib] [PySpark] Python KMeans parity

### PR DESCRIPTION
**Pull request by FlytxtRnD**
The following items are added to Python kmeans:

kmeans - setEpsilon, setInitializationSteps
KMeansModel - computeCost, k

### CODE REVIEW

**Comment by FlytxtRnD** [comment]
Could anyone please tell us why this timeout has occurred?

**Comment by FlytxtRnD** [comment]
Jenkins, retest this please

**Comment by jkbradley** [comment]
@FlytxtRnD Checking in: What's the status of fixing the tests?  Thanks!  It will be great to get this merged for 1.4 (deadline this Friday!)

Btw, the earlier test timeout was from Jenkins having a bunch of issues, but hopefully those are resolved now.

**Comment by FlytxtRnD** [comment]
@jkbradley, we are facing some issues with python 3 support. We are working on it and will fix it asap.

File: python/pyspark/mllib/clustering.py [hunk 1]
```diff
@@ -95,6 +106,13 @@ def predict(self, x):
                 best_distance = distance
         return best
 
+    def computeCost(self, rdd):
+        """Return the K-means cost (sum of squared distances of points to their nearest center) for this
```
Comment (jkbradley) [inline python/pyspark/mllib/clustering.py]: python style (newline for long doc strings + shorter doc lines <= 72 chars in length following PEP 8):
```
"""
Return the K-means cost (sum of squared distances of points to their
nearest center) for this model on the given data.
"""
```

**Comment by jkbradley** [comment]
@FlytxtRnD That tiny issue is the only one I see.  After that, I think this PR will be good to go.  Thanks!

File: python/pyspark/mllib/clustering.py [hunk 2]
```diff
@@ -40,11 +40,16 @@ class KMeansModel(Saveable, Loader):
 
     >>> data = array([0.0,0.0, 1.0,1.0, 9.0,8.0, 8.0,9.0]).reshape(4, 2)
     >>> model = KMeans.train(
-    ...     sc.parallelize(data), 2, maxIterations=10, runs=30, initializationMode="random")
+    ...     sc.parallelize(data), 2, maxIterations=10, runs=30, initializationMode="random",
+    ...                    seed=None, initializationSteps=5, epsilon=1e-4)
```
Comment (jkbradley) [inline python/pyspark/mllib/clustering.py]: I'm sorry about this, but hoped to add 1 more update.  Could you please set the seed to some fixed number in this and the other call to train() in this doc test?  It should be deterministic for stability.  (It looks like this was a problem before your PR too.)  Thanks!

File: python/pyspark/mllib/clustering.py [hunk 3]
```diff
@@ -52,7 +57,8 @@ class KMeansModel(Saveable, Loader):
     ...     SparseVector(3, {2: 1.0}),
     ...     SparseVector(3, {2: 1.1})
     ... ]
-    >>> model = KMeans.train(sc.parallelize(sparse_data), 2, initializationMode="k-means||")
+    >>> model = KMeans.train(sc.parallelize(sparse_data), 2, initializationMode="k-means||",
```
Comment (jkbradley) [inline python/pyspark/mllib/clustering.py]: here too

**Comment by jkbradley** [comment]
After the seed update, it really will be it.  Thanks!

File: python/pyspark/mllib/clustering.py [hunk 4]
```diff
@@ -40,11 +40,16 @@ class KMeansModel(Saveable, Loader):
...
+    ...                    seed=None, initializationSteps=5, epsilon=1e-4)
     >>> model.predict(array([0.0, 0.0])) == model.predict(array([1.0, 1.0]))
     True
     >>> model.predict(array([8.0, 9.0])) == model.predict(array([9.0, 8.0]))
     True
+    >>> model.k
+    2
+    >>> model.computeCost(sc.parallelize(data))
+    2.0000000000000004
     >>> model = KMeans.train(sc.parallelize(data), 2)
```
Comment (FlytxtRnD) [inline python/pyspark/mllib/clustering.py]: @jkbradley , It seems we are not using this model anywhere. Did you mean to add the seed here too?

Comment (jkbradley) [inline python/pyspark/mllib/clustering.py]: If it's not used anywhere, then you can leave it as is.  Thanks!

Comment (FlytxtRnD) [inline python/pyspark/mllib/clustering.py]: Shall we remove that line?

**Comment by FlytxtRnD** [comment]
@jkbradley , could you please check if this is ready to merge?

File: python/pyspark/mllib/clustering.py (hunk 4 above)
Comment (jkbradley) [inline python/pyspark/mllib/clustering.py]: No, let's keep it

**Comment by jkbradley** [comment]
LGTM.  I'm rerunning tests once more for safety sake and then will merge this.  Thank you!

**Comment by jkbradley** [comment]
unrelated failure...

**Comment by jkbradley** [comment]
@FlytxtRnD The update confused github.  Can you please try closing and re-opening this PR to force github to recompute the diff?  Thanks!

**Comment by FlytxtRnD** [comment]
@jkbradley ok, will close and reopen this.  Could you please tell us why this happened?

**Comment by jkbradley** [comment]
Hm, that didn't seem to fix it.  It looks like you merged this branch with the current master; sometimes, that confuses Github.  In general, you shouldn't need to merge with master unless Jenkins posts a notice that the PR can't be merged cleanly.

Can you please try rebasing your branch off of the current master?  Perhaps that will fix it.

**Comment by FlytxtRnD** [comment]
@jkbradley , we merged it with branch-1.4 to get rid of the failed Mima tests in jenkins.

**Comment by FlytxtRnD** [comment]
"Can you please try rebasing your branch off of the current master? Perhaps that will fix it."
We didn't get your point.. To which branch should we rebase to?

**Comment by jkbradley** [comment]
Looking at those logs, I don't think the MIMA tests actually failed.  Basing this off of "master" should be fine.  Could you please try that?

If that does not work, then perhaps you could identify your commit hashes, reset this branch to master,  cherry pick your commits, and then force push to your remote branch to update this PR.

**Comment by FlytxtRnD** [comment]
@jkbradley , ok we'll try your steps.

**Comment by FlytxtRnD** [comment]
Jenkins, retest this please

**Comment by mengxr** [comment]
Merged into master and branch-1.4. Thanks!

**Comment by FlytxtRnD** [comment]
@jkbradley , @mengxr  Thanks for the help!

**Pull request closed without merging**